crates/node/src/failure_detector/fd_state.rs

```diff
@@ -268,6 +279,29 @@ impl FdState {
         }
     }
 
+    pub fn update_from_cluster_state_message(
+        &mut self,
+        opts: &GossipOptions,
+        cs_reply: ClusterStateReply,
+    ) {
+        for incoming_node in cs_reply.nodes {
+            if incoming_node.node_id == self.my_node_id {
+                // We already know our own state, no need to update it.
+                continue;
+            }
+
+            if incoming_node.state == net::node::NodeState::Alive {
+                let node = self.get_node_or_insert(incoming_node.node_id);
+                // reset the age of the node if it's alive
+                node.gossip_age = 0;
+                node.state = NodeState::Alive;
+                self.cs_updater
+                    .upsert_node_state(incoming_node.node_id, NodeState::Alive.into());
+            }
+        }
+        debug!("{}", self.dump_as_string(opts));
```

Comment:
Should we also update the replica set states based on the `partitions` field of `cs_reply`?

Reply:
Yes. Good call. I was waiting for your PR in case it changed anything in replicaset states API, so I'll rebase once yours is merged.

Reply:
I've updated mine, this should be ready to merge, hopefully no changes here that conflict with your PR.